### PR TITLE
WIFI-1228 get clients by macAddress

### PR DESCRIPTION
**Pull request by ralphlee3**

### CODE REVIEW

File: client-datastore-inmemory/src/main/java/com/telecominfraproject/wlan/client/datastore/inmemory/ClientDatastoreInMemory.java [hunk 1]
```diff
@@ -176,6 +176,120 @@ public Client delete(int customerId, MacAddress clientMac) {
...
+        	//this datastore is only interested in the last item's id, so we'll clear all other fields on the startAfterItem in the pagination context
+        	Client newStartAfterItem = new Client();
+        	newStartAfterItem.setCustomerId(ret.getContext().getStartAfterItem().getCustomerId());
+        	newStartAfterItem.setMacAddress(ret.getContext().getStartAfterItem().getMacAddress());
+        	ret.getContext().setStartAfterItem(newStartAfterItem);
+        }
+
+        return ret;
+
+//    	List<Client> ret = new ArrayList<>();
```
Comment (dmitry-toptygin-connectus-ai) [inline client-datastore-inmemory/src/main/java/com/telecominfraproject/wlan/client/datastore/inmemory/ClientDatastoreInMemory.java:277]: please remove this old code

File: client-datastore-inmemory/src/main/java/com/telecominfraproject/wlan/client/datastore/inmemory/ClientDatastoreInMemory.java [hunk 2]
```diff
@@ -176,6 +176,120 @@ public Client delete(int customerId, MacAddress clientMac) {
...
+        ret.setContext(context.clone());
+
+        if (ret.getContext().isLastPage()) {
+            // no more pages available according to the context
+            return ret;
+        }
+
+        List<Client> items = new LinkedList<>();
+        
+    	if(macSubstring != null) {
```
Comment (dmitry-toptygin-connectus-ai) [inline client-datastore-inmemory/src/main/java/com/telecominfraproject/wlan/client/datastore/inmemory/ClientDatastoreInMemory.java:198]: This is the only null check for the macSubstring in the whole PR.
Usually null or empty criteria means that caller wants to see all items.
The way you implemented it right now for the null criteria - InMemory datastore will return no items, RDBMS and Cassandra datastores will throw NullPointerException.
Please align the datastores on common behaviour (I suggest to return all items), and add unit tests for the null and "" macSubstring criteria.

File: client-datastore-cassandra/src/main/java/com/telecominfraproject/wlan/client/datastore/cassandra/ClientDAO.java [hunk 3]
```diff
@@ -142,7 +142,11 @@
     private static final String CQL_DELETE_BLOCKED_LIST = 
     		"delete from client_blocklist where customerId = ? and  macAddress = ?";
     
-    private static final String CQL_GET_BY_MAC_STRING = 
+    // Statements using client_by_mac_string table
+    private static final String CQL_GET_CLIENT_MAC_BY_CUSTOMER_ID = 
+    		"select customerId, macAddress from client_by_mac_string where customerId = ? and macAddressString like ?";
```
Comment (dmitry-toptygin-connectus-ai) [inline client-datastore-cassandra/src/main/java/com/telecominfraproject/wlan/client/datastore/cassandra/ClientDAO.java:147]: nope. you probably meant just 
"select customerId, macAddress from client_by_mac_string where customerId = ?";

File: client-datastore-cassandra/src/main/java/com/telecominfraproject/wlan/client/datastore/cassandra/ClientDAO.java [hunk 4]
```diff
@@ -378,6 +411,91 @@ public Client delete(int customerId, MacAddress clientMac) {
...
+		ResultSet rs = cqlSession.execute(boundStmt);
+		ByteBuffer nextPagingState = rs.getExecutionInfo().getPagingState();
+
+		List<Client> pageItems = new ArrayList<>();
+		
+		// iterate through the current page
+		while (rs.getAvailableWithoutFetching() > 0) {
+			Set<MacAddress> macSet = new HashSet<>();
+			rs.all().forEach(y -> macSet.add(new MacAddress(y.getLong("macAddress"))));
+			pageItems.addAll(get(customerId, macSet));
```
Comment (dmitry-toptygin-connectus-ai) [inline client-datastore-cassandra/src/main/java/com/telecominfraproject/wlan/client/datastore/cassandra/ClientDAO.java:471]: rs.all() returns all items in the query regardless of the pagination constraints. you need to use rs.one() in the loop with rs.getAvailableWithoutFetching(). and pageItems.addAll() should happen outside that loop.

File: client-datastore-cassandra/src/main/java/com/telecominfraproject/wlan/client/datastore/cassandra/ClientDAO.java [hunk 5]
```diff
@@ -378,6 +411,91 @@ public Client delete(int customerId, MacAddress clientMac) {
...
+        // ***** We will ignore the order supplied by the caller for this datastore
+        
+        ArrayList<Object> bindVars = new ArrayList<>();
+        BoundStatement boundStmt;
+        
+        if (macSubstring != null) {
+		    bindVars.add(customerId);
+		    bindVars.add("%" + macSubstring.toLowerCase() + "%");
+	                
+	        boundStmt = preparedStmt_getClientMacLikeMacString.bind(bindVars.toArray());
```
Comment (dmitry-toptygin-connectus-ai) [inline client-datastore-cassandra/src/main/java/com/telecominfraproject/wlan/client/datastore/cassandra/ClientDAO.java:447]: this line could go below the if() - it happens in both if and else

File: client-datastore-cassandra/src/main/java/com/telecominfraproject/wlan/client/datastore/cassandra/ClientDAO.java [hunk 6]
```diff
@@ -378,6 +411,91 @@ public Client delete(int customerId, MacAddress clientMac) {
...
+        } else {
+		    bindVars.add(customerId);
+	                
+	        boundStmt = preparedStmt_getClientMacByCustomerId.bind(bindVars.toArray());
```
Comment (dmitry-toptygin-connectus-ai) [inline client-datastore-cassandra/src/main/java/com/telecominfraproject/wlan/client/datastore/cassandra/ClientDAO.java:451]: this line could go below the if() - it happens in both if and else

Comment (ralphlee3) [reply client-datastore-cassandra/src/main/java/com/telecominfraproject/wlan/client/datastore/cassandra/ClientDAO.java:451]: The 2 lines are using different PreparedStatements, one without the substring condition and one with the substring. 

Did you mean to get the proper statement in the if() and then bind them outside the if()?

Comment (dmitry-toptygin-connectus-ai) [reply client-datastore-cassandra/src/main/java/com/telecominfraproject/wlan/client/datastore/cassandra/ClientDAO.java:451]: You are right, those are two different statements, and they should stay as they are. 
Sorry, my bad.

**Review by dmitry-toptygin-connectus-ai — APPROVED** [review]
